vizro-core/src/vizro/managers/_data_manager.py

```diff
@@ -45,14 +65,30 @@ def __setitem__(self, dataset_name: DatasetName, data: Union[pd.DataFrame, pd_La
             raise ValueError(f"Dataset {dataset_name} already exists.")
 
         if callable(data):
-            self.__lazy_data[dataset_name] = data
+            data = VizroDataSet(data)
         elif isinstance(data, pd.DataFrame):
             self.__original_data[dataset_name] = data
+        elif isinstance(data, VizroDataSet):
+            pass
         else:
             raise TypeError(
                 f"Dataset {dataset_name} must be a pandas DataFrame or callable that returns pandas DataFrame."
+                f"Additionally, it can be a VizroDataSet object."
             )
 
+        if isinstance(data, VizroDataSet):
+            self.__lazy_data[dataset_name] = data
+
+    def __getitem__(self, dataset_name: DatasetName) -> VizroDataSet:
+        """Returns the `VizroDataSet` object associated with `dataset_name`."""
+        if dataset_name not in self.__original_data and dataset_name not in self.__lazy_data:
+            raise KeyError(f"Dataset {dataset_name} does not exist.")
+        if dataset_name in self.__original_data:
+            # no cache available
+            raise ValueError(f"Dataset {dataset_name} is not lazy.")
```

Comment:
I'm not sure I like these switches around original data vs. lazy data but let's leave like this for now and then we can possibly change to Idea 2 after we merge this into #133. That will simplify things quite a bit I think.